code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ProductCatalogNotificationBuilderTest.kt

```diff
@@ -0,0 +1,147 @@
+    @Mock
+    private lateinit var context: Context
+    private lateinit var trackerActivityClass: Class<out Activity>
+    private lateinit var broadcastReceiverClass: Class<out BroadcastReceiver>
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
+
+        context = RuntimeEnvironment.getApplication()
+        trackerActivityClass = DummyActivity::class.java
+        broadcastReceiverClass = DummyBroadcastReceiver::class.java
+        mockkObject(PushTemplateImageUtils)
+    }
+
+    @Test
+    fun `construct should throw NotificationConstructionFailedException if downloaded image count does not match catalog image count`() {
+        val pushTemplate = provideMockedProductCatalogTemplate()
+
+        assertFailsWith(
+            exceptionClass = NotificationConstructionFailedException::class,
+            message = "Failed to download all images for the product catalog notification.",
+            block = {
+                ProductCatalogNotificationBuilder.construct(context, pushTemplate, trackerActivityClass, broadcastReceiverClass)
+            }
+        )
+    }
+
+    @Test
+    fun `construct should return a NotificationCompat Builder if download image count matches catalog image count`() {
+        val pushTemplate = provideMockedProductCatalogTemplate()
+        val cachedItem = mockkClass(Bitmap::class)
+
+        // product catalog template requires 3 catalog items
+        every { cacheImages(any()) } answers { 3 }
+        every { getCachedImage(any()) } answers { cachedItem }
+        val notificationBuilder = ProductCatalogNotificationBuilder.construct(context, pushTemplate, trackerActivityClass, broadcastReceiverClass)
+
+        assertEquals(R.layout.push_template_horizontal_catalog, notificationBuilder.bigContentView.layoutId)
+        assertEquals(NotificationCompat.Builder::class.java, notificationBuilder::class.java)
+    }
+
+    @Test
+    fun `construct should build a notification with a vertical layout if vertical layout is provided`() {
+        val dataMap = getMockedMapWithProductCatalogData()
+        dataMap.replaceValueInMap(PushTemplateConstants.PushPayloadKeys.CATALOG_LAYOUT, PRODUCT_CATALOG_VERTICAL_LAYOUT)
+        val pushTemplate = ProductCatalogPushTemplate(MapData(dataMap))
+        val cachedItem = mockkClass(Bitmap::class)
+
+        // product catalog template requires 3 catalog items
+        every { cacheImages(any()) } answers { 3 }
+        every { getCachedImage(any()) } answers { cachedItem }
+
+        val notificationBuilder = ProductCatalogNotificationBuilder.construct(context, pushTemplate, trackerActivityClass, broadcastReceiverClass)
+        assertEquals(R.layout.push_tempate_vertical_catalog, notificationBuilder.bigContentView.layoutId)
+        assertEquals(NotificationCompat.Builder::class.java, notificationBuilder::class.java)
+    }
+
+    @Test
+    fun `construct should return a valid NotificationCompat Builder Broadcast Receiver is not provided`() {
```

Comment:
nit: construct should return a valid NotificationCompat Builder if the Broadcast Receiver is not provided